godot-core/src/storage.rs

```diff
@@ -109,40 +102,46 @@ mod single_thread {
             })
         }
 
+        pub fn base(&self) -> &Base<T::Base> {
+            &self.cached_base
+        }
+
         pub(super) fn godot_ref_count(&self) -> u32 {
-            self.godot_ref_count
+            self.godot_ref_count.get()
         }
     }
 }
 
 #[cfg(feature = "threads")]
 mod multi_thread {
     use std::any::type_name;
-    use std::sync;
     use std::sync::atomic::{AtomicU32, Ordering};
+    use std::{cell, sync};
 
-    use crate::obj::GodotClass;
-    use crate::out;
+    use crate::obj::{Base, GodotClass};
+    use crate::{out, sys};
 
     use super::Lifecycle;
 
     /// Manages storage and lifecycle of user's extension class instances.
     pub struct InstanceStorage<T: GodotClass> {
         user_instance: sync::RwLock<T>,
+        cached_base: Base<T::Base>,
 
         // Declared after `user_instance`, is dropped last
-        pub lifecycle: Lifecycle,
+        pub lifecycle: cell::Cell<Lifecycle>,
```

Comment:
`Cell` isn't sync, so i dont think this is safe. we could make this an atomic integer instead.

Reply:
There's a cool trick for these things, btw:
```rs
fn enforce_sync<T: Sync>();

// never called
fn __static_type_check() {
    enforce_sync::<Type>();
}
```

I need to see if it's already applicable, if not I'd like to avoid spending time now to improve multi-threaded soundness, this needs more thought anyway...

Reply:
I cannot apply this yet. Just discovered that `T` would need to be `Sync` and likely `Send` if used in a thread-shared `Gd<T>` under the `threads` feature. This will however mean adding those bounds in a ton of places, which I'm not gonna do now.

Possibly we could imply the bounds in something like
```
trait GodotClass: 'static + Sync + Send {
    ...
}
```
but then that would also affect engine types, not just user-defined `T`. Which I guess is the right move, anyway.

Reply:
one possible solution to this is to have a sync version of `Gd`. which can only hold `Sync` rust objects (would we need `Send`? i guess the destructor takes the object out of the `Gd`, and i assume it can be called on another thread.)

which does sorta mirror how the std has sync and unsync versions of some types. like `Rc` vs `Arc`.

Reply:
> would we need `Send`? i guess the destructor takes the object out of the `Gd`, and i assume it can be called on another thread

Not just the destructor, but as long as `bind_mut()` still hands out a guard with `DerefMut`, one can do this:
```rs
// on thread 1:
let obj = MyClass { ... };
let ptr = Gd::new(obj);

// on thread 2:
let ptr: Gd<MyClass> // passed from thread 1
let guard = ptr.bind_mut();

let extracted = mem::take(*guard); // replaces with default
// do whatever on thread 2
```
`Sync` works with `&T` access + interior mutability, but not really with `&mut` because that is essentially a free pass to get the value...

---

> it also means we dont need a separate feature for thread safety, we'd just need people to use the right smart pointer.

This whole topic probably needs more thought... In general, I'd also like to avoid _accidental_ use of threads when there's no explicit opt-in, i.e. runtime validations in some places.

Having two types can also be unergonomic, requiring extra conversions, and less flexible API design. On GDScript there is no such distinction, so there may be friction at the boundary. Another option is also special `impl <T: GodotClass + Sync> Gd<T> { ... }` blocks provide functionality on top.